File: designer/src/com/android/tools/idea/uibuilder/surface/CanvasResizeLayer.java
package com.android.tools.idea.uibuilder.surface;

import static com.android.tools.idea.uibuilder.graphics.NlConstants.BOUNDS_RECT_DELTA;
import static com.android.tools.idea.uibuilder.graphics.NlConstants.RESIZING_CORNER_COLOR;
import static com.android.tools.idea.uibuilder.graphics.NlConstants.RESIZING_CUE_COLOR;
import static com.android.tools.idea.uibuilder.graphics.NlConstants.RESIZING_HOVERING_SIZE;

import com.android.tools.adtui.common.SwingCoordinate;
import com.android.tools.idea.common.surface.Layer;
import com.android.tools.idea.uibuilder.graphics.NlConstants;
import java.awt.Dimension;
import java.awt.Graphics2D;
import java.awt.Rectangle;
import org.jetbrains.annotations.NotNull;

/**
 * Layer to buildDisplayList the canvas resizing cue in the bottom-right corner of the screen view.
 */
public class CanvasResizeLayer extends Layer {
  private final NlDesignSurface myDesignSurface;
  private final ScreenView myScreenView;
  private boolean myIsHovering;

  public CanvasResizeLayer(@NotNull NlDesignSurface designSurface, @NotNull ScreenView screenView) {
    myDesignSurface = designSurface;
    myScreenView = screenView;
  }

  /**
   * Sets the state of this layer according to the mouse hovering at point (x, y).
   * Returns whether that required any modification to the state of the layer.
   */
  @Override
  public void onHover(@SwingCoordinate int x, @SwingCoordinate int y) {
    boolean oldHovering = myIsHovering;
    Dimension size = myScreenView.getScaledContentSize();
    Rectangle resizeZone =
      new Rectangle(myScreenView.getX() + size.width, myScreenView.getY() + size.height, RESIZING_HOVERING_SIZE, RESIZING_HOVERING_SIZE);
    myIsHovering = resizeZone.contains(x, y);
    if (myIsHovering != oldHovering) {
      myDesignSurface.repaint();
    }
  }

  @Override
  public void paint(@NotNull Graphics2D g2d) {
    Dimension size = myScreenView.getScaledContentSize();
    int x = myScreenView.getX();
    int y = myScreenView.getY();

    Graphics2D graphics = (Graphics2D)g2d.create();
    graphics.setStroke(NlConstants.SOLID_STROKE);
    graphics.setColor(myIsHovering ? RESIZING_CORNER_COLOR : RESIZING_CUE_COLOR);
    graphics.drawLine(x + size.width + BOUNDS_RECT_DELTA, y + size.height + 4, x + size.width + 4, y + size.height + BOUNDS_RECT_DELTA);
    graphics.drawLine(x + size.width + BOUNDS_RECT_DELTA, y + size.height + 12, x + size.width + 12, y + size.height + BOUNDS_RECT_DELTA);
    graphics.dispose();
  }

  @Override
  public boolean isVisible() {
    return myScreenView.getScene().isResizeAvailable();
  }
}

File: designer/src/com/android/tools/idea/uibuilder/surface/layer/CanvasResizeLayer.kt
/*
 * Copyright (C) 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.uibuilder.surface.layer

import com.android.tools.adtui.common.SwingCoordinate
import com.android.tools.idea.common.surface.Layer
import com.android.tools.idea.uibuilder.graphics.NlConstants
import com.android.tools.idea.uibuilder.surface.ScreenView
import java.awt.Graphics2D
import java.awt.Rectangle

/**
 * Layer to buildDisplayList the canvas resizing cue in the bottom-right corner of the screen view.
 */
class CanvasResizeLayer(private val screenView: ScreenView, private val repaint: () -> Unit) :
  Layer() {
  private var isHovering = false

  /**
   * Sets the state of this layer according to the mouse hovering at point (x, y). Returns whether
   * that required any modification to the state of the layer.
   */
  override fun onHover(@SwingCoordinate x: Int, @SwingCoordinate y: Int) {
    val oldHovering = isHovering
    val size = screenView.scaledContentSize
    val resizeZone =
      Rectangle(
        screenView.x + size.width,
        screenView.y + size.height,
        NlConstants.RESIZING_HOVERING_SIZE,
        NlConstants.RESIZING_HOVERING_SIZE,
      )
    isHovering = resizeZone.contains(x, y)
    if (isHovering != oldHovering) {
      repaint()
    }
  }

  override fun paint(g2d: Graphics2D) {
    val size = screenView.scaledContentSize
    val x = screenView.x
    val y = screenView.y
    val graphics = g2d.create() as Graphics2D
    graphics.stroke = NlConstants.SOLID_STROKE
    graphics.color =
      if (isHovering) NlConstants.RESIZING_CORNER_COLOR else NlConstants.RESIZING_CUE_COLOR
    graphics.drawLine(
      x + size.width + NlConstants.BOUNDS_RECT_DELTA,
      y + size.height + 4,
      x + size.width + 4,
      y + size.height + NlConstants.BOUNDS_RECT_DELTA,
    )
    graphics.drawLine(
      x + size.width + NlConstants.BOUNDS_RECT_DELTA,
      y + size.height + 12,
      x + size.width + 12,
      y + size.height + NlConstants.BOUNDS_RECT_DELTA,
    )
    graphics.dispose()
  }

  override val isVisible: Boolean
    get() = screenView.scene.isResizeAvailable
}
